[BugFix] Fix race condition when access delete predicate list in tablet meta (#42099)

srlch opened the pull request:
## Why I'm doing:
tablet::version_for_delete_predicate do not hold lock to access delete predicate list which may cause crash.

## What I'm doing:
Try to acquire lock in share mode when access delete predicate list

Fixes #42099

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [x] 3.0
  - [x] 2.5

imay commented on be/src/storage/tablet.cpp:
```diff
@@ -836,9 +836,19 @@ Status Tablet::_capture_consistent_rowsets_unlocked(const std::vector<Version>&
+    // this the current thread in shared/exclusive mode. It may
+    // be crash if we acquire share lock again. We just do nothing
+    // in this case.
+    try {
+        std::shared_lock rlock(get_header_lock());
+    } catch (const std::exception& e) {}
```
This lock will be released after this try block, are you sure that this will fix the bug?

srlch replied:
Yes, it is a problem, maybe we can use get_header_lock().lock_shared() instead. But i still reviewing implementation is reasonable or not. I will fix it later

srlch replied:
updated

srlch committed:
[BugFix] Fix race condition when access delete predicate list in tablet meta (#42099)

Why I'm doing:
tablet::version_for_delete_predicate do not hold lock to access delete predicate list
which may cause crash.

What I'm doing:
separate tablet::version_for_delete_predicate into to function:
tablet::version_for_delete_predicate
tablet::version_for_delete_predicate_unlocked
Use this two functions base on the caller has acquire the _meta_lock or not

Signed-off-by: srlch <linzichao@starrocks.com>

trueeyu commented on be/src/storage/tablet.cpp:
```diff
@@ -836,9 +838,12 @@ Status Tablet::_capture_consistent_rowsets_unlocked(const std::vector<Version>&
     return Status::OK();
 }
```
Tablet::get_compaction_status() also need use version_for_delete_predicate_unlocked?

srlch replied:
updated

srlch committed:
Fix

Signed-off-by: srlch <linzichao@starrocks.com>

srlch committed:
Fix

Signed-off-by: srlch <linzichao@starrocks.com>

srlch committed:
Fix

Signed-off-by: srlch <linzichao@starrocks.com>
